app/src/components/eventStreamService.js

```diff
@@ -0,0 +1,333 @@
+const falsey = require('falsey');
+const { JSONCodec } = require('nats');
+
+// different connection libraries if we are using websockerts or nats protocols.
+const WEBSOCKETS = !falsey(config.get('eventStreamService.websockets'));
+
+let natsConnect;
+if (WEBSOCKETS) {
+  // shim the websocket library
+  globalThis.WebSocket = require('websocket').w3cwebsocket;
+  const { connect } = require('nats.ws');
+  natsConnect = connect;
+} else {
+  const { connect } = require('nats');
+  natsConnect = connect;
+}
+
+const log = require('./log')(module.filename);
+
+const { FormVersion, Form, FormEventStreamConfig } = require('../forms/common/models');
+const formMetadataService = require('../forms/form/formMetadata/service');
+const { featureFlags } = require('./featureFlags');
+const { encryptionService } = require('./encryptionService');
+
+const SERVICE = 'EventStreamService';
+
+const FORM_EVENT_TYPES = {
+  PUBLISHED: 'published',
+  UNPUBLISHED: 'unpublished',
+};
+
+const SUBMISSION_EVENT_TYPES = {
+  CREATED: 'created',
+  UPDATED: 'updated',
+  DELETED: 'deleted',
+};
+
+const jsonCodec = JSONCodec();
+
+class DummyEventStreamService {
+  // this class should not be called if we actually check that this feature is enabled.
+  // however... if we missed that check these calls will do nothing.
+  async checkConnection() {
+    log.warn('EventStreamService.checkConnection - EventStreamService is not enabled.');
+  }
+  async openConnection() {
+    log.warn('EventStreamService.openConnection - EventStreamService is not enabled.');
+  }
+  closeConnection() {
+    log.warn('EventStreamService.closeConnection - EventStreamService is not enabled.');
+  }
+  get connected() {
+    log.warn('EventStreamService.connected - EventStreamService is not enabled.');
+    return false;
+  }
+  // eslint-disable-next-line no-unused-vars
+  async onPublish(formId, formVersionId, published) {
+    log.warn('EventStreamService.onPublish - EventStreamService is not enabled.');
+  }
+  // eslint-disable-next-line no-unused-vars
+  async onSubmit(eventType, submission, draft) {
+    log.warn('EventStreamService.onSubmit - EventStreamService is not enabled.');
+  }
+}
+
+class EventStreamService {
+  constructor({ servers, streamName, source, domain, username, password }) {
+    if (!servers || !streamName || !source || !domain || !username || !password) {
+      throw new Error('EventStreamService is not configured. Check configuration.');
+    }
+
+    this.servers = servers;
+    this.streamName = streamName;
+    this.source = source;
+    this.domain = domain;
+    this.username = username;
+    this.password = password;
+
+    this.nc = null; // nats connection
+    this.js = null; // jet stream
+    this.jsm = null; // jet stream manager
+    this.natsOptions = {
+      servers: this.servers.split(','),
+      maxReconnectAttempts: 1,
+      name: this.streamName,
+      reconnectTimeWait: 5000, // wait 5 seconds before retrying...
```

Comment:
It looks like on line 87 it's configured to wait 5s before reconnecting, but on line 85 there's only one reconnect - does this mean that it does indeed do a single retry if NATS is unreachable? Would it be more resilient to retry more than once? I guess the bigger question is if it is always asynchronous to send to NATS, and if so then maybe a bunch of retries might be a good thing?

Reply:
not being an expert on NATS, just using some reasonable values for that connection library. i believe I cribbed those values from the connection used in the education apps?
here's more info if you want to read, I don't know what the best values to set are.
https://docs.nats.io/using-nats/developer/connecting/reconnect#reconnection-attributes